BES_DrawTrace: Get rid of member variable mCurrentNetSignal

5n8ke opened the pull request:
Fix crash when trying to connect to a pad that is not part of a signal.

Fixes #762

ubruhin commented:
Btw, #753 will introduce the same issue again just a few lines below. Maybe better move the exception to outside the if..else if..else if construct to make sure *any* case of a missing netsignal is catched?

5n8ke commented:
Actually the assert could be replaced with the exception.

But I want to take a closer look at mCurrentNetSignal, since it's only set when accessing a device pad and not for the other items.

5n8ke commented:
mCurrentNetSignal is now removed since the correct netsignal can always be gotten from mCurrentNetSegment.
I introduced the same check in #753 where it definitely caused crashes and would update this PR as soon as the other one is merged.

5n8ke commented:
Since #753 is merged, I'll update this PR tonight.

ubruhin commented:
> Since #753 is merged, I'll update this PR tonight.

Perfect! Afterwards I'll rebase and merge #760 if there are no objections.

5n8ke commented:
Not on my side.

5n8ke commented:
@ubruhin It's updated to the current master version

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_drawtrace.cpp:
```diff
@@ -918,23 +916,27 @@ void BES_DrawTrace::updateNetpointPositions() noexcept {
     Q_ASSERT(layer);
-    if (BI_Via* via =
-            findVia(board, mCursorPos, mCurrentNetSignal, {mTempVia})) {
+    NetSignal* netsignal = &mCurrentNetSegment->getNetSignal();
+    // NOTE(5n8ke): netsignal must not be nullptr, since a connection should
+    // only be made to the current NetSignal
+    Q_ASSERT(netsignal);
```
`netsignal` can't be `nullptr` since it was constructed by taking the address of an object, so no check is needed:

```suggestion
    NetSignal* netsignal = &mCurrentNetSegment->getNetSignal();
```

But actually, one more time, simply don't convert the reference to a pointer here, then it's obvious that it can't be `nullptr` :smiley: :

```suggestion
    NetSignal& netsignal = mCurrentNetSegment->getNetSignal();
```

Then use `&netsignal` where you really need a pointer.

5n8ke replied:
I chose to use a pointer here since all the following find* take a pointer as the netsignal parameter.
The check is not needed, but I added the asserts as a sanity check to make it explicit, since a null pointer was previously possible with mCurrentNetSignal.

ubruhin replied:
Hmm okay - to be honest, I don't like it (IMO a reference would be much cleaner) but this time I'll turn a blind eye :wink:

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_drawtrace.cpp:
```diff
@@ -976,11 +978,14 @@ void BES_DrawTrace::layerComboBoxIndexChanged(int index) noexcept {
-    if (findVia(board, startPos, mCurrentNetSignal) ||
+    Board&     board     = mPositioningNetPoint1->getBoard();
+    Point      startPos  = mFixedStartAnchor->getPosition();
+    NetSignal* netsignal = &mCurrentNetSegment->getNetSignal();
+    // NOTE(5n8ke): netsignal must not be nullptr, since a connection should
+    // only be made to the current NetSignal
+    Q_ASSERT(netsignal);
```
Same here:

```suggestion
    NetSignal& netsignal = mCurrentNetSegment->getNetSignal();
```

ubruhin reviewed (commented):
Although it turned out that the "fix" is actually not needed (sorry...), I always like removing members since usually it reduces complexity and error-proneness :smiley:
